content/posts/jspi-and-function-pointer-cast-handling.md

```diff
@@ -0,0 +1,380 @@
+    <span class="nb">table.get</span> <span class="nv">$functable</span>
+    <span class="nb">local.tee</span> <span class="nv">$funcref</span>
+    <span class="c1"></span>
+    <span class="c1">;; Two args?</span>
+    <span class="nb">ref.test</span> <span class="p">(</span><span class="k">ref</span> <span class="nv">$two_args</span><span class="p">)</span>
+    <span class="k">if</span>
+      <span class="nb">i32.const</span> <span class="mi">2</span>
+      <span class="nb">return</span>
+    <span class="k">end</span>
+    <span class="nb">local.get</span> <span class="nv">$funcref</span>
+    <span class="c1"></span>
+    <span class="c1">;; One arg?</span>
+    <span class="nb">ref.test</span> <span class="p">(</span><span class="k">ref</span> <span class="nv">$one_args</span><span class="p">)</span>
+    <span class="k">if</span>
+      <span class="nb">i32.const</span> <span class="mi">1</span>
+      <span class="nb">return</span>
+    <span class="k">end</span>
+    <span class="nb">local.get</span> <span class="nv">$funcref</span>
+    <span class="c1"></span>
+    <span class="c1">;; Zero arg?</span>
+    <span class="nb">ref.test</span> <span class="p">(</span><span class="k">ref</span> <span class="nv">$zero_args</span><span class="p">)</span>
+    <span class="k">if</span>
+      <span class="nb">i32.const</span> <span class="mi">0</span>
+      <span class="nb">return</span>
+    <span class="k">end</span>
+    <span class="c1"></span>
+    <span class="c1">;; It takes more than two args, or uses a non-i32 type.</span>
+    <span class="nb">i32.const</span> <span class="mi">-1</span>
+    <span class="nb">return</span>
+  <span class="p">)</span>
+<span class="p">)</span>
+  </code></pre>
+</div>
+
+Now we can use this as follows:
+```C
+typedef int (*F)(int, int);
+typedef int (*F1)(int);
+typedef int (*F0)(void);
+
+WASM_IMPORT("countArgs")
+int countArgs(F func);
+
+int callHandler(F f, int x, int y) {
+  int nargs = countArgs(f);
+  switch (nargs) {
+    case 2:  
+      logString("Two arguments");
+      return func(x, y);
+    case 1:
+      logString("One argument");
+      return ((F1)func)(x);
+    case 0:
+      logString("Zero arguments");
+      return ((F0)func)();
+    default:
+      logString("Bad handler");
+      return -1;
+  }
+}
+
+WASM_EXPORT("fakePyFunc")
+void fakePyFunc(int func_index, int x, int y) {
+  int res = callHandler(handlers[func_index], x, y);
+
+  logString("Got result:");
+  logInt(res);
+}
+```
+We can use `wasm-as` to convert the wat to wasm and use `wasm-merge` to merge
+`countArgs.wasm` with the compiled C code. `wasm-as` and `wasm-merge` are part
+of binaryen. See the full build script here:
+[here](https://github.com/hoodmane/jspi-blog-examples/tree/main/9-wat-count-args/build.sh).
+
+This approach to handling function pointer casts has the disadvantage that it
+requires wasm-gc which has only been supported in Safari since December of 2024.
```

Comment:
I am curious why wasm-gc is a hard requirement for this. Is it because `ref.test` is part of wasm-gc?